File: frameworks/juniper/tests/mutations_test.rs
mod test_helpers;
use test_helpers::*;

// ============================================================================
// Mutation: updateUser
// ============================================================================

#[test]
fn test_update_user_full_name() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.full_name = "Alice Smith".to_string();

    // Verify
    assert_eq!(user.full_name, "Alice Smith");
    assert_eq!(user.id, user_id);
}

#[test]
fn test_update_user_bio() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.bio = Some("Senior Developer".to_string());

    // Verify
    assert_eq!(user.bio, Some("Senior Developer".to_string()));
    assert_eq!(user.id, user_id);
}

#[test]
fn test_update_user_both_fields() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.full_name = "Alice Smith".to_string();
    user.bio = Some("Senior Developer".to_string());

    // Verify
    assert_eq!(user.full_name, "Alice Smith");
    assert_eq!(user.bio, Some("Senior Developer".to_string()));
    assert_eq!(user.id, user_id);
}

#[test]
fn test_update_user_clear_bio() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.bio = None;

    // Verify
    assert_eq!(user.bio, None);
    assert_eq!(user.id, user_id);
}

// ============================================================================
// Mutation: updatePost
// ============================================================================

#[test]
fn test_update_post_title() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Original Title", "Original Content");
    let post_id = post.id.clone();

    // Simulate mutation
    post.title = "Updated Title".to_string();

    // Verify
    assert_eq!(post.title, "Updated Title");
    assert_eq!(post.id, post_id);
}

#[test]
fn test_update_post_content() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Original Title", "Original Content");
    let post_id = post.id.clone();

    // Simulate mutation
    post.content = "Updated Content".to_string();

    // Verify
    assert_eq!(post.content, "Updated Content");
    assert_eq!(post.id, post_id);
}

#[test]
fn test_update_post_both_fields() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Original Title", "Original Content");
    let post_id = post.id.clone();

    // Simulate mutation
    post.title = "Updated Title".to_string();
    post.content = "Updated Content".to_string();

    // Verify
    assert_eq!(post.title, "Updated Title");
    assert_eq!(post.content, "Updated Content");
    assert_eq!(post.id, post_id);
}

// ============================================================================
// Mutation: Field Immutability
// ============================================================================

#[test]
fn test_user_id_immutable_after_update() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Bio");
    let original_id = user.id.clone();

    // Try to "update"
    user.bio = Some("Updated".to_string());

    // Verify ID unchanged
    assert_eq!(user.id, original_id);
}

#[test]
fn test_post_id_immutable_after_update() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Title", "Content");
    let original_id = post.id.clone();

    // Try to "update"
    post.title = "Updated".to_string();

    // Verify ID unchanged
    assert_eq!(post.id, original_id);
}

#[test]
fn test_username_immutable() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let original_username = user.username.clone();

    // Try to "update"
    user.bio = Some("Updated".to_string());

    // Verify username unchanged
    assert_eq!(user.username, original_username);
}

// ============================================================================
// Mutation: State Changes
// ============================================================================

#[test]
fn test_sequential_updates_accumulate() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "");

    // Apply updates sequentially
    user.bio = Some("Developer".to_string());
    user.bio = Some("Senior Developer".to_string());

    // Verify latest state
    assert_eq!(user.bio, Some("Senior Developer".to_string()));
}

#[test]
fn test_updates_isolated_between_entities() {
    let factory = TestFactory::new();
    let mut user1 = factory.create_test_user("alice", "alice@example.com", "Alice", "Bio1");
    let mut user2 = factory.create_test_user("bob", "bob@example.com", "Bob", "Bio2");

    let original_bio2 = user2.bio.clone();

    // Update user1
    user1.bio = Some("Updated".to_string());

    // Verify user2 unchanged
    assert_eq!(user2.bio, original_bio2);
}

// ============================================================================
// Mutation: Return Value Validation
// ============================================================================

#[test]
fn test_updated_user_returns_all_fields() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    user.bio = Some("Updated".to_string());

    // Verify all fields present
    assert!(!user.id.is_empty());
    assert_eq!(user.username, "alice");
    assert_eq!(user.bio, Some("Updated".to_string()));
}

#[test]
fn test_updated_post_returns_all_fields() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Title", "Content");
    post.title = "Updated".to_string();

    // Verify all fields present
    assert!(!post.id.is_empty());
    assert_eq!(post.title, "Updated");
    assert!(post.fk_author > 0);
}

#[test]
fn test_mutation_maintains_created_at() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let original_created_at = user.created_at.clone();

    // Update
    user.full_name = "Alice Updated".to_string();

    // Verify created_at unchanged
    assert_eq!(user.created_at, original_created_at);
}

File: frameworks/csharp-dotnet/FraiseQL.Benchmark.Tests/MutationsTest.cs
using FraiseQL.Benchmark.Models;
using Xunit;
using System;

namespace FraiseQL.Benchmark.Tests;

public class MutationsTest
{
    private readonly TestFactory _factory = new();

    // ============================================================================
    // Mutation: updateUser
    // ============================================================================

    [Fact]
    public void TestUpdateUserFullName()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Developer");
        var userId = user.Id;

        // Simulate mutation
        user.FullName = "Alice Smith";

        // Verify
        Assert.Equal("Alice Smith", user.FullName);
        Assert.Equal(userId, user.Id);
    }

    [Fact]
    public void TestUpdateUserBio()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Developer");
        var userId = user.Id;

        // Simulate mutation
        user.Bio = "Senior Developer";

        // Verify
        Assert.Equal("Senior Developer", user.Bio);
        Assert.Equal(userId, user.Id);
    }

    [Fact]
    public void TestUpdateUserBothFields()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Developer");
        var userId = user.Id;

        // Simulate mutation
        user.FullName = "Alice Smith";
        user.Bio = "Senior Developer";

        // Verify
        Assert.Equal("Alice Smith", user.FullName);
        Assert.Equal("Senior Developer", user.Bio);
        Assert.Equal(userId, user.Id);
    }

    [Fact]
    public void TestUpdateUserClearBio()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Developer");
        var userId = user.Id;

        // Simulate mutation
        user.Bio = null;

        // Verify
        Assert.Null(user.Bio);
        Assert.Equal(userId, user.Id);
    }

    // ============================================================================
    // Mutation: updatePost
    // ============================================================================

    [Fact]
    public void TestUpdatePostTitle()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Original Title", "Original Content");
        var postId = post.Id;

        // Simulate mutation
        post.Title = "Updated Title";

        // Verify
        Assert.Equal("Updated Title", post.Title);
        Assert.Equal(postId, post.Id);
    }

    [Fact]
    public void TestUpdatePostContent()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Original Title", "Original Content");
        var postId = post.Id;

        // Simulate mutation
        post.Content = "Updated Content";

        // Verify
        Assert.Equal("Updated Content", post.Content);
        Assert.Equal(postId, post.Id);
    }

    [Fact]
    public void TestUpdatePostBothFields()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Original Title", "Original Content");
        var postId = post.Id;

        // Simulate mutation
        post.Title = "Updated Title";
        post.Content = "Updated Content";

        // Verify
        Assert.Equal("Updated Title", post.Title);
        Assert.Equal("Updated Content", post.Content);
        Assert.Equal(postId, post.Id);
    }

    // ============================================================================
    // Field Immutability
    // ============================================================================

    [Fact]
    public void TestUserIdImmutableAfterUpdate()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Bio");
        var originalId = user.Id;

        // Try to "update"
        user.Bio = "Updated";

        // Verify ID unchanged
        Assert.Equal(originalId, user.Id);
    }

    [Fact]
    public void TestPostIdImmutableAfterUpdate()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Title", "Content");
        var originalId = post.Id;

        // Try to "update"
        post.Title = "Updated";

        // Verify ID unchanged
        Assert.Equal(originalId, post.Id);
    }

    [Fact]
    public void TestUsernameImmutable()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        var originalUsername = user.Username;

        // Try to "update"
        user.Bio = "Updated";

        // Verify username unchanged
        Assert.Equal(originalUsername, user.Username);
    }

    // ============================================================================
    // State Changes
    // ============================================================================

    [Fact]
    public void TestSequentialUpdatesAccumulate()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");

        // Apply updates sequentially
        user.Bio = "Developer";
        user.Bio = "Senior Developer";

        // Verify latest state
        Assert.Equal("Senior Developer", user.Bio);
    }

    [Fact]
    public void TestUpdatesIsolatedBetweenEntities()
    {
        var user1 = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Bio1");
        var user2 = _factory.CreateTestUser("bob", "bob@example.com", "Bob", "Bio2");

        var originalBio2 = user2.Bio;

        // Update user1
        user1.Bio = "Updated";

        // Verify user2 unchanged
        Assert.Equal(originalBio2, user2.Bio);
    }

    // ============================================================================
    // Return Value Validation
    // ============================================================================

    [Fact]
    public void TestUpdatedUserReturnsAllFields()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Developer");
        user.Bio = "Updated";

        // Verify all fields present
        Assert.NotEqual(Guid.Empty, user.Id);
        Assert.Equal("alice", user.Username);
        Assert.Equal("Updated", user.Bio);
    }

    [Fact]
    public void TestUpdatedPostReturnsAllFields()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Title", "Content");
        post.Title = "Updated";

        // Verify all fields present
        Assert.NotEqual(Guid.Empty, post.Id);
        Assert.Equal("Updated", post.Title);
        Assert.NotNull(post.Author);
    }

    [Fact]
    public void TestMutationMaintainsCreatedAt()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        var originalCreatedAt = user.CreatedAt;

        // Update
        user.FullName = "Alice Updated";

        // Verify created_at unchanged
        Assert.Equal(originalCreatedAt, user.CreatedAt);
    }
}
